### ISSUE TITLE
Track provenance/fixity correctly

### ISSUE THREAD

**Issue by sellout** [ndmitchell/hlint#1088]
I've seen badly-typed hints when hlint isn't aware of the fixity of an operator. E.g., suggesting
```haskell
foo . bar =<< f <$> g <*\> h
```
(where `<*\>` is effectively identical to `<*>`) be rewritten to
```haskell
foo . bar . f =<< g <*\> h
```
which is not well-typed. It seems there are already a couple other issues here that reference specific cases of this, like #883 and #980. Hopefully those could be closed as duplicates of this task to fix the underlying issue.

I _think_ hlint effectively needs to run the typechecker (or at least use more of the GHC API) to be able to load everything that's in scope and get the correct types and fixities. Fixity seems to already be handled correctly for locally-defined things (e.g., if `<*\>` has a fixity declaration in the same module as the expression above, hlint does the right thing), so loading all the dependencies should make it feasible to use that same logic globally.

**Comment by ndmitchell** [comment]
From https://github.com/ndmitchell/hlint#bugs-and-limitations:

> HLint operates on each module at a time in isolation, as a result HLint does not know about types or which names are in scope.

This is very deliberate. Using the GHC API introduces a whole heap of additional failure modes and compatibility constraints. It also reduces parallelism. HLint does not intend to work with types, or type checked modules, or dependencies - it's just too complex to orchestrate reliably.

However, HLint as integrated into IDE's (e.g. haskell-language-server) will be supplied with a list of the fixities in scope, which should make it give the right answer with respect to fixities. 

For your specific program you can also list the fixities HLint knows about with: https://github.com/ndmitchell/hlint#why-doesnt-hlint-know-the-fixity-for-my-custom--operator

**Comment by sellout** [comment]
Ah, thanks for the explanation. I'm aware of the hlint fixity declarations, but I considered them a workaround until now. I think capturing the motivation in the docs would be helpful, as I read that as just describing the current state rather than an explicit decision (especially since until recently hlint wasn't wasn't using the GHC API).

But this makes sense, thanks again.

**Comment by ndmitchell** [comment]
Good idea - I've added:

> HLint operates on each module at a time in isolation, as a result HLint does not know about types or which names are in scope. This decision is deliberate, allowing HLint to parallelise and be used incrementally on code that may not type-check. If fixities are required to parse the code properly, they [can be supplied](https://github.com/ndmitchell/hlint#why-doesnt-hlint-know-the-fixity-for-my-custom--operator).

Does that capture it for you?
